ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -294,3 +324,116 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+    -- ^ There may multiple ids for one expression.
+    -- e.g. @(a,b) = (1,True)@
+    , bindingLoc :: SrcSpan
+    -- ^ Location for the whole binding.
+    -- Here we use the this to render the type signature at the proper place.
+    --
+    -- Example: For @(a,b) = (1,True)@, it will print the signature after the
+    -- open parenthesis instead of the above of the whole expression.
+    }
+
+-- | Existed bindings in a where clause.
+data WhereBindings = WhereBindings
+    { bindings        :: [WhereBinding]
+    , existedSigNames :: [Name]
+    -- ^ Names of existing signatures.
+    -- It is used to hide type lens for existing signatures.
+    }
+
+-- | All where clauses from type checked source.
+findWhereQ :: GenericQ [LHsLocalBinds GhcTc]
+findWhereQ = everything (<>) $ mkQ [] (pure . findWhere)
+    where
+        findWhere :: GRHSs GhcTc (LHsExpr GhcTc) -> LHsLocalBinds GhcTc
+        findWhere = grhssLocalBinds
+
+-- | Find all bindings for **one** where clasure.
+findBindingsQ :: GenericQ (Maybe WhereBindings)
+findBindingsQ = something (mkQ Nothing findBindings)
+    where
+        findBindings :: NHsValBindsLR GhcTc -> Maybe WhereBindings
+        findBindings (NValBinds binds sigs) =
+            Just $ WhereBindings
+                { bindings = mapMaybe (something (mkQ Nothing findBindingIds) . snd) binds
+                , existedSigNames = concatMap findSigIds sigs
+                }
+
+        findBindingIds :: LHsBindLR GhcTc GhcTc -> Maybe WhereBinding
+        findBindingIds (L l FunBind{..}) = Just $ WhereBinding (pure $ unLoc fun_id) l
+        findBindingIds (L l PatBind{..}) =
+            let ids = (everything (<>) $ mkQ [] (maybeToList . findIdFromPat)) pat_lhs
+            in Just $ WhereBinding ids l
+        findBindingIds _ = Nothing
+
+        -- | Example: Find `a` and `b` from @(a,b) = (1,True)@
+        findIdFromPat :: Pat GhcTc -> Maybe Id
+        findIdFromPat (VarPat _ (L _ id)) = Just id
+        findIdFromPat _                   = Nothing
+
+        findSigIds (L _ (TypeSig _ names _)) = map unLoc names
+        findSigIds _                         = []
+
+-- | Provide code lens for where bindings.
+whereClauseCodeLens :: PluginMethodHandler IdeState TextDocumentCodeLens
+whereClauseCodeLens state plId CodeLensParams{..} = do
+    enabled <- usePropertyLsp #whereLensOn plId properties
+    if not enabled then pure $ pure $ List [] else pluginResponse $ do
+        nfp <- getNormalizedFilePath plId uri
+        tmr <- handleMaybeM "Unable to typechecking"
+            $ liftIO
+            $ runAction "codeLens.local.TypeCheck" state
+            $ use TypeCheck nfp
+        (hscEnv -> hsc) <- handleMaybeM "Unable to get GhcSession"
+            $ liftIO
+            $ runAction "codeLens.local.GhcSession" state
+            $ use GhcSession nfp
+        let tcGblEnv = tmrTypechecked tmr
+            rdrEnv = tcg_rdr_env tcGblEnv
+            typeCheckedSource = tcg_binds tcGblEnv
```

Comment:
I wonder if it's worth trying to merge this with the other code that generates the lenses so we only do this stuff once. Not sure how expensive it is. But then maybe it's hard to manage them being enabled/disabled :thinking: